src/app/clusters/network-commissioning/ThreadScanResponse.cpp

```diff
@@ -0,0 +1,146 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "ThreadScanResponse.h"
+
+#include "constants.h"
+#include <clusters/NetworkCommissioning/Commands.h>
+#include <clusters/NetworkCommissioning/Structs.h>
+#include <lib/support/SortUtils.h>
+
+using chip::DeviceLayer::NetworkCommissioning::ThreadScanResponse;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace NetworkCommissioning {
+
+namespace {
+
+/// Fills up scanResponseArray with valid and de-duplicated thread responses from mNetworks.
+/// Handles sorting and keeping only larger rssi
+///
+/// Returns the valid list of scan responses into `validResponses`, which is only valid
+/// as long as scanResponseArray is valid.
+CHIP_ERROR LoadResponses(DeviceLayer::NetworkCommissioning::ThreadScanResponseIterator * networks,
+                         Platform::ScopedMemoryBuffer<ThreadScanResponse> & scanResponseArray,
+                         Span<ThreadScanResponse> & validResponses)
+{
+    VerifyOrReturnError(scanResponseArray.Alloc(std::min(networks == nullptr ? 0 : networks->Count(), kMaxNetworksInScanResponse)),
+                        CHIP_ERROR_NO_MEMORY);
+
+    ThreadScanResponse scanResponse;
+    size_t scanResponseArrayLength = 0;
+    for (; networks != nullptr && networks->Next(scanResponse);)
+    {
+        if ((scanResponseArrayLength == kMaxNetworksInScanResponse) &&
+            (scanResponseArray[scanResponseArrayLength - 1].rssi > scanResponse.rssi))
+        {
+            continue;
+        }
+
+        bool isDuplicated = false;
+
+        for (size_t i = 0; i < scanResponseArrayLength; i++)
+        {
+            if ((scanResponseArray[i].panId == scanResponse.panId) &&
+                (scanResponseArray[i].extendedPanId == scanResponse.extendedPanId))
+            {
+                if (scanResponseArray[i].rssi < scanResponse.rssi)
+                {
+                    scanResponseArray[i] = scanResponseArray[--scanResponseArrayLength];
+                }
+                else
+                {
+                    isDuplicated = true;
+                }
+                break;
+            }
+        }
+
+        if (isDuplicated)
+        {
+            continue;
+        }
+
+        if (scanResponseArrayLength < kMaxNetworksInScanResponse)
+        {
+            scanResponseArrayLength++;
+        }
+        scanResponseArray[scanResponseArrayLength - 1] = scanResponse;
+
+        // TODO: this is a sort (insertion sort even, so O(n^2)) in a O(n) loop.
+        ///      There should be some better alternatives to not have some O(n^3) processing complexity.
+        Sorting::InsertionSort(scanResponseArray.Get(), scanResponseArrayLength,
```

Comment:
Why is the sorting not done once at the end? In terms of keeping only strongest RSSI, we could easily do it by first checking if RSSI is is >= the min of all current RSSI (O(N) worst case) so we would be at O(N^2) worst case rather than O(N^3).

Reply:
This would need review on actual intent because that does not seem to be  commented.
My guess is that it tries to be space efficient while keeping the strongest RSSI ... but it does so at a large expense of O). 

TO refactor this we would need some good test cases. Now we can add some for sure, however even that is a larger tast.

Reply:
created #39301